README
SfS Goodies for R --- Useful R functions from Seminar fuer Statistik, ETH Zurich


ChangeLog
2013-01-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/ps.goodies.R (pdf.do): set paper to "special" when user
	specifies 'width' and 'height'.

2013-01-03  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-24,

	* R/prettylab.R (eaxis): for 'log': new 'between.max = 4';
	set 'at.small <- FALSE' in such cases, when small ticks are not
	sensible.


2012-11-01  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-23, released to CRAN on 2012-11-01

	* R/misc-goodies.R (digitsBase): ndigits default argument needs
	fuzz (1e-9), e.g., for base 3.

2012-10-20  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-22, released to CRAN on 2012-10-20

	* R/ps.goodies.R (dev.latex): oops: finally found and fixed bug
	where the missing(.) checks where never true, as dev.latex() is
	always called from above with explicit argument passing.

	* man/histBxp.Rd: renamed hist.bxp() to histBxp() .. on CRAN's
	urging.

2012-09-26  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-21, released to CRAN today
	* R/prettylab.R (eaxis): new 'small.args' for Marius

2012-04-18  Martin Maechler  <maechler@stat.math.ethz.ch>

	* man/unix/Sys.cpuinfo.Rd: fix example

2012-03-18  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-20, released to CRAN on 2012-03-18

	* man/capture-n-write.Rd: new capture.and.write() function

2011-11-21  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/misc-goodies.R (seqXtend): change to enable "interpolate" with
	'Date' objects.

2011-10-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-19, released to CRAN on 2011-11-21

	* man/str_data.Rd: + examples for 'filterFUN'

2011-10-08  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/str_data.R (str_data): new arg 'filterFUN'

2011-10-03  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-18, released to CRAN.
	* R/prettylab.R (eaxis): fix Rversion check.

2011-10-01  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/prettylab.R (pretty10exp): digits.fuzz = 7:
	add fuzz before rounding, twice.
	(eaxis): new arg  draw.between.ticks = TRUE.

2011-07-29  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-17, released to CRAN on 2011-10-01.

	* R/sourceAttach.R (sourceAttach): new utility function

2011-07-27  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-16, released to CRAN.

	* R/pd-matrix.R (posdefify): symmetric = TRUE is now default, as
	that's needed for asymmetric input.

	* R/misc-goodies.R: get completely rid of boxplot.matrix: that's
	been in R, since 2.9.0 now.

2011-05-27  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-15, released to CRAN.

	* R/misc-goodies.R (errbar): 'ylim' now is argument (with same default).

2011-05-04  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/prime-numbers-fn.R: finally prime number and factorization
	"utility" code (partly ~ 13 years old!) to a place where it's
	easily found.

2011-04-28  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/prettylab.R (eaxis): new argument 'max.at'

	* man/eaxis.Rd: fix docu for at.small & small.mult.

2011-04-18  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/diagDA.R (predict.dDA): check for var()==0 als for pool=FALSE.

2010-12-06  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-14

	* R/pd-matrix.R (posdefify): new argument 'eigen.m'; for efficiency
	to pass to eigen() in case it's already available.

2010-11-04  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/misc-goodies.R (inv.seq): workaround mapply() infelicity

2010-10-27  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/misc-goodies.R (repChar): new utility {generalization of bl.string()}

	* R/printTable.R (margin2table): return *named* dimnames, if
	original has them.
	(print.margin2table): use 'right = TRUE' by default

2010-10-20  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-13;  "Enhances:" for everything in our examples

	* inst/NEWS: partial update

2010-09-28  Martin Maechler  <maechler@stat.math.ethz.ch>

	* demo/hatmat-ex.R: add demo for hatMat(), notably how it works for
	loess(); rather than extending the examples in
	* man/hatMat.Rd

2010-09-04  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/ps.goodies.R (ps.end): call .set.eps_view() only when needed.

	* R/zzz.R (.set.eps_view): utility function instead of using
	.onLoad() for setting options("eps_view").


2010-08-28  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-12

	* R/misc-goodies.R (plotDS): fix bug for unsorted x and "moreSmooth"

2010-02-21  Martin Maechler  <maechler@stat.math.ethz.ch>

	* man/nearcor.Rd: cor() no longer works for factors;
	--> adapt the old example (from Jens or other users).

2009-12-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* man/mat2tex.Rd: need quadruple escapes ( \\\\pi )

2009-12-14  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-10

	* R/ps.goodies.R (pdf.do): paper="default", width=-1, height=-1
	such as to produce the same default as ps.do().

	* man/ps.latex.Rd: update accordingly

2009-11-18  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-9, CRAN-released: today

	* man/polyn.eval.Rd, man/p.ts.R, man/eaxis.Rd: fix \link{}s

2009-10-07  Martin Maechler  <maechler@stat.math.ethz.ch>

	* R/misc-goodies.R (lseq): check that 'from > 0'

2009-08-10  Martin Maechler  <maechler@stat.math.ethz.ch>

	* DESCRIPTION (Version): 1.0-8

	* R/unix/Sys.ps.R (Sys.procinfo): produce unique names;
	important for multi-core/processor CPUs.

	* man/unix/Sys.cpuinfo.Rd: ditto

2009-07-16  Martin Maechler  <maechler@stat.math.ethz.ch>

	* man/uniqueL.Rd: link to Duplicated

2009-04-16  Martin Maechler  <maechler@stat.math.ethz.ch>


R/TA.plot.R
#### $Id: TA.plot.R,v 1.11 2012/03/18 21:53:39 maechler Exp $
n.plot <-
    function(x, y=NULL, nam = NULL, abbr = n >= 20 || max(nchar(nam))>=8,
             xlab = NULL, ylab = NULL, log = "",
             cex = par("cex"), col = par("col"), ...)
{
    ## Purpose: "Name Plot"; Names (or numbers) instead of points in plot(..)
    ## --> help(n.plot) !
    if(inherits(x,"formula")) # is(x, "formula")
        stop("formula not yet supported")
    ## this is like plot.default():
    xlabel <- if (!missing(x)) deparse(substitute(x))
    ylabel <- if (!missing(y)) deparse(substitute(y))
    xy <- xy.coords(x, y, xlabel, ylabel, log)
    xlab <- if (is.null(xlab)) xy$xlab else xlab
    ylab <- if (is.null(ylab)) xy$ylab else ylab
    plot(xy, type = 'n', xlab = xlab, ylab = ylab, log = log, ...)
    n <- length(x)
    if(is.null(nam)) {  nam <- rownames(x)
     if (is.null(nam)) { nam <- names(x)
      if (is.null(nam)) { nam <- names(y)
       if (is.null(nam)) { nam <- paste(1:n) #- Use 1,2,.. if no names
    }}}}
    if(abbr) nam <- abbreviate(nam, minlength=1)
    text(xy, labels=nam, cex=cex, col=col)
    invisible(nam)
}

TA.plot <-
  function(lm.res, fit = fitted(lm.res),
           res = residuals(lm.res, type = "pearson"),
           labels = NULL, main = mk.main(), xlab = "Fitted values",
           draw.smooth = n >= 10, show.call = TRUE, show.2sigma = TRUE,
           lo.iter = NULL, lo.cex = NULL,
           par0line  = list(lty = 2, col = "gray"),
           parSmooth = list(lwd = 1.5, lty = 4, col = 2),
           parSigma  = list(lwd = 1.2, lty = 3, col = 4),
           ...)
{
  ## Purpose: Produce a Tukey-Anscombe plot of a linear model fit
  ##	      Note that residuals and fitted are UN-correlated (IFF intercept..)
  ## -------------------------------------------------------------------------
  ## Arguments: lm.res = RESult of lm(..)
  ##    res : (weighted) residuals by default,
  ##	labels = 'symbols' for point, default(NULL): extract names or use seq.nr
  ##             use '*' to get simple '*' symbols.
  ##
  ## --- see on-line help by  "?TA.plot" !!
  ## -------------------------------------------------------------------------
  ## Uses : n.plot(.)
  ## -------------------------------------------------------------------------
  ## Author: Martin Maechler, Date: Dec 92 / Nov.93;  for R: 1999/2000
  if(missing(main)) {
    call <- lm.res $ call
    if(is.call(call[["formula"]]) && any(c("lm", "aov") == call[[1]]))
      call <- call[["formula"]]
    else {  #-- only formula part; no extra  'data ='
        if (length(call) >= 3 && !is.na(m.f <- match("formula", names(call)))) {
            call <- call[c(1, m.f)]
            names(call)[2] <- ""
        }
    }
    mk.main <- function() {
      cal <- call ## if(is.R()) call else get("call", frame = sys.parent())
      if(is.null(cal))
        "Tukey-Anscombe plot of ???"
      else {
	  nc <- nchar(ccal <- deparse(cal, width.cutoff = 200)[1])
	  prt.DEBUG("|cal|=", length(cal), "; nchar(ccal) =", nc,": '", ccal,
		    "'\n", sep="")
	  if(nc > 36)
	      warning("TA.plot: 'main' title is long; consider using cex.main = 0.8",
		      call. = FALSE)
	  ##-- now should even go further:
	  ##--  E.g. if  nc > 50,  use  cex = .8 in the call to n.plot below
	  paste(if(nc < 13) "Tukey-Anscombe plot of :  "
		else if(nc < 24) "T.A. plot of: " else "TA-pl:", ccal)
	}
    }
  }
  if("ylab" %in% names(list(...))) {
      n.plot(fit, res, nam = labels, xlab = xlab, main = main, ...)
  } else {
      yl <- "Residuals"
      if(!is.null(lm.res$weights) &&
         any(abs(lm.res$resid- res) > 1e-6*mad(res)))
          yl <- paste("WEIGHTED", yl)
      n.plot(fit, res, nam = labels, xlab = xlab, ylab = yl, main = main, ...)
  }
  if(show.call)
    mtext(deparse(match.call()), side = 3, line = 0.5, cex = 0.6, adj=1)
  do.call("abline", c(list(h= 0), par0line))
  p.mgp <- par("mgp")[1:2] #-- line numbers of margin text: xlab & label
  if(missing(lo.cex))
    lo.cex <- max(.2, min(0.8*par("mex"), .9*-diff(p.mgp))/par("mfg")[4])
  m.line <- if(par("mfg")[4]==1) .2+ p.mgp[1] else
                              max(p.mgp[1] - .2*lo.cex, sum(p.mgp)/2)
  if(show.2sigma) {
    s2 <- c(-2,2) * mad(res, center=0)
    rr <- range(res)
    if(s2[1] < rr[1] || s2[2] > rr[2])
      mtext(paste("2 sigma = ", format(s2[2])),
	    side= 1, line= m.line, adj = 0, cex= lo.cex)
    ##abline(h= s2, lwd=1.8, lty=3, col=4)
    do.call("abline", c(list(h= s2), parSigma))
  }
  n <- length(res)
  if(draw.smooth) {
    if(!is.list(parSmooth)) stop("`parSmooth' must be a list")
    ##-- lo.iter: idea of Werner Stahel:  no robustness for 'glm'  residuals
    if (is.null(lo.iter))
      lo.iter <- if(inherits(lm.res, "glm")&& lm.res$family[1]!="Gaussian")
	0  else  3
    f <- max(0.2, 1.25 * n^-.2) #'-- Martin's very empirical formula...
    rlow <- lowess(fit, res, f = f, iter = lo.iter)
    do.call("lines",c(rlow, parSmooth))

    mtext(paste("-.-.-.- : lowess smooth (f =", format(round(f,2)),
		if(lo.iter!=3) paste(", it=", lo.iter), ")"),
	  side = 1, line = m.line, cex = lo.cex, adj = 1)
  }
  ##-	 "Correlation:", formatC(cor(fit,res), dig=3),
  ## mtext(paste(" -- Rank corr.:", formatC(cor(rank(fit),rank(res)), dig=3)) )
  invisible()
}
